components/style_traits/values/font.rs

```diff
@@ -0,0 +1,379 @@
+        match *self {
+            SingleFontFamily::FamilyName(ref name) => name.to_css(dest),
+
+            // All generic values accepted by the parser are known to not require escaping.
+            SingleFontFamily::Generic(ref name) => {
+                #[cfg(feature = "gecko")] {
+                    // We should treat -moz-fixed as monospace
+                    if name == &atom!("-moz-fixed") {
+                        return dest.write_str("monospace");
+                    }
+                }
+
+                write!(dest, "{}", name)
+            },
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, MallocSizeOf, PartialEq)]
+#[cfg_attr(feature = "servo", derive(Deserialize, Serialize))]
+/// The name of a font family of choice
+pub struct FamilyName {
+    /// Name of the font family
+    pub name: Atom,
+    /// Syntax of the font family
+    pub syntax: FamilyNameSyntax,
+}
+
+impl ToCss for FamilyName {
+    fn to_css<W>(&self, dest: &mut CssWriter<W>) -> fmt::Result where W: fmt::Write {
+        match self.syntax {
+            FamilyNameSyntax::Quoted => {
+                dest.write_char('"')?;
+                write!(CssStringWriter::new(dest), "{}", self.name)?;
+                dest.write_char('"')
+            }
+            FamilyNameSyntax::Identifiers(ref serialization) => {
+                // Note that `serialization` is already escaped/
+                // serialized appropriately.
+                dest.write_str(&*serialization)
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, MallocSizeOf, PartialEq)]
+#[cfg_attr(feature = "servo", derive(Deserialize, Serialize))]
+/// Font family names must either be given quoted as strings,
+/// or unquoted as a sequence of one or more identifiers.
+pub enum FamilyNameSyntax {
+    /// The family name was specified in a quoted form, e.g. "Font Name"
+    /// or 'Font Name'.
+    Quoted,
+
+    /// The family name was specified in an unquoted form as a sequence of
+    /// identifiers.  The `String` is the serialization of the sequence of
+    /// identifiers.
+    Identifiers(String),
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, MallocSizeOf, PartialEq, Serialize)]
+pub enum FontStretch {
+    Normal,
+    UltraCondensed,
+    ExtraCondensed,
+    Condensed,
+    SemiCondensed,
+    SemiExpanded,
+    Expanded,
+    ExtraExpanded,
+    UltraExpanded,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, MallocSizeOf, PartialEq, Serialize)]
+pub enum FontVariantCaps {
+    Normal,
+    SmallCaps,
+}
+
+/// Everything gfx needs from style::style_structs::Font
+pub trait FontStyleStruct {
+    fn get_size(&self) -> Au;
+    fn get_hash(&self) -> u64;
+    fn get_font_weight(&self) -> FontWeight;
+    fn get_font_stretch(&self) -> FontStretch;
+    fn get_font_variant_caps(&self) -> FontVariantCaps;
+    fn get_font_families(&self) -> slice::Iter<SingleFontFamily>;
```

Comment:
Given `gfx` only cares about font family names, what about doing something that doesn't require to move all the font-family code into `style_traits`, and instead just adds something like:

```rust
fn each_family_name<F>(&self, f: F)
where
    F: FnMut(&str);
```

Reply:
This will allow fixing the `cfg(feature = "gecko")` build errors.

Reply:
Yeah that seem nicer, I wonder if it would also be better to do something similar to the `style::font_face::Source` stuff. Like passing one function for the `Url` case and one for the `Local` case.